File: src/optical_element/MATLAB/beam.m
classdef beam <handle 
    %波动光学中的光束    
    properties   
        wavelength;     % 波长，单位           [m]
        X;              % 网格坐标 X方向
        Y;              % 网格坐标 Y方向
        field;          % 光束场 复数 二维矩阵        
        N;              % 网格的长度 单位 点的个数
        phase_field;    % 相位场分布
        XX;%              散点化
         YY;%散点化
%         phi;psi;n;
     
    end
    
    methods
        function obj = beam(X,Y,field,wavelength)
            obj.X =X;
            obj.Y =Y;
            obj.field =field;
            obj.wavelength =wavelength; 
            obj.N=size(X,1);                            %网格坐标X（横向，沿着行方向）方向的维度数
            %obj.path();
        end
    end    
    
   methods  % plot 强度分布图和相位分布图  输入 distance为距离 单位 [m]
        function  detector_field(obj,distance)                  
            addpath(genpath('.\display'));                            %添加 显示二维强度函数的路径
             if (nargin<2)                                            %如果参数小于2个，直接 作图
                  Furion_intensity_detector(obj.X,obj.Y,obj.field,0); %作图函数 输入 X 网格 Y 网格 field 复数场 
                  Furion_phase_detector(obj.X,obj.Y,obj.field,2);  
             else
                  bs=obj.translation(distance);                        %如果2个参数，先将光线传播一段距离 distance，
                  Furion_intensity_detector(bs.X,bs.Y,bs.field,0);     %然后作图显示强度分布
                  Furion_phase_detector(obj.X,obj.Y,obj.field,1);      %作图显示相位分布
             end          
        end
   end  
      
   methods % 光束线传播一段距离   distance 距离 单位[m]  
        function  beam_tanslated = translation(obj,distance)       
             flag = 2;%选择传播子
             L = max(obj.X(1,:))-min(obj.X(1,:));%空间采样的尺寸
             if flag ==1                 
                 out_Field = Furion_fresnel_spatialh (obj.X,obj.Y,obj.field,obj.wavelength,distance); 
                 %菲涅尔传播子、实域空间采样
             elseif flag ==2
                 out_Field = Furion_fresnel_spectralh (obj.N,1/L,obj.field,obj.wavelength,distance);
                 %菲涅尔传播子，频域空间采样
             elseif flag == 3
                 out_Field = Furion_angular_spectrum (obj.N,1/L,obj.field,obj.wavelength,distance);
             else
                 out_Field = fresnel_propagation_mono_2D(obj.field,L,L,obj.wavelength,distance) ;
             end
                  beam_tanslated = beam(obj.X,obj.Y,out_Field,obj.wavelength);     %生成新的光束       
        end
   end
       
    methods%重采样  %适用于平面镜、柱面镜  %重新采样 L重新采样的尺寸大小 单位 [m]
        function  beam_=resize1(obj,L)   
           x = linspace(-L/2,L/2,obj.N);                                        %重新采样x轴方向
            y = x';                                                             %重新采样Y轴方向
            [X_r,Y_r] = meshgrid(x,y);   
            %创建重新采样的网格点
 
            
            F = scatteredInterpolant(obj.X(:),obj.Y(:),obj.field(:),'linear');  %散点插值  一维数组
            Field_new = F(X_r(:),Y_r(:));                                       %插值得到新的采样
            Field_new_1 =reshape(Field_new,obj.N,obj.N);                        %矩阵一维转换为二维
           beam_ = beam(X_r,Y_r,Field_new_1,obj.wavelength);                    %创建新的光束         
        end
    end
    
     methods%重采样  %适用于平面镜、柱面镜  %重新采样 L重新采样的尺寸大小 单位 [m]
        function  beam_=resize2(obj,L,N)   %   改变size和抽样点数
           x = linspace(-L/2,L/2,N);                                        %重新采样x轴方向
            y = x';                                                             %重新采样Y轴方向
            [X_r,Y_r] = meshgrid(x,y);   
            %创建重新采样的网格点
 
            
            F = scatteredInterpolant(obj.X(:),obj.Y(:),obj.field(:),'linear');  %散点插值  一维数组
            Field_new = F(X_r(:),Y_r(:));                                       %插值得到新的采样
            Field_new_1 =reshape(Field_new,N,N);                        %矩阵一维转换为二维
           beam_ = beam(X_r,Y_r,Field_new_1,obj.wavelength);                    %创建新的光束         
        end
     end
    
         methods%重采样  %适用于平面镜、柱面镜  %重新采样 L重新采样的尺寸大小 单位 [m]
        function  beam_=resize3(obj,Lx, Ly,N)   %   改变size和抽样点数
           x = linspace(-Lx/2,Lx/2,N);                                        %重新采样x轴方向
            y = linspace(-Ly/2, Ly/2, N)';                                                             %重新采样Y轴方向
            [X_r,Y_r] = meshgrid(x,y);   
            %创建重新采样的网格点
 
            
            F = interp2(obj.X,obj.Y,obj.field,X_r,Y_r,'cubic',0);  %散点插值  一维数组
            %Field_new = F(X_r(:),Y_r(:));                                       %插值得到新的采样
            %Field_new_1 =reshape(Field_new,N,N);                        %矩阵一维转换为二维
           beam_ = beam(X_r,Y_r,F,obj.wavelength);                    %创建新的光束         
        end
    end
  
methods       %计算光束的总功率
    function p = power(obj)
          x = obj.X(1,:);
          y = obj.Y(:,1);
          p = trapz(y,trapz(x,abs(obj.field).^2,2));
        % p= sum(sum(abs(obj.field).^2))*dx*dy;%振幅的平方求和
    end
end

end

File: src/optical_element/beam.cpp
#include "Beam.h"

using namespace Furion_NS;

Beam::Beam(std::vector<std::vector<double> >& X, std::vector<std::vector<double> >& Y, std::vector<std::vector<std::complex<double>> >& field, double wavelength, int N) :
    wavelength(wavelength), X(X), Y(Y), field(field), N(N)//, XX(nullptr), YY(nullptr), phase_field(nullptr)
{
    cout << "Beam的初始化" << endl;
    //create_2d(phase_field, N);
    //create_2d(kk, N);
    this->kk = std::vector<std::vector<double> >(N, std::vector<double>(N));
}

Beam::~Beam()
{
    //destory_2d(X, N);
    //destory_2d(Y, N);
    //destory_2d(field, N);
    //destory_1d(XX);
    //destory_1d(YY);
    cout << "Beam的析构" << endl;
}

void Beam::destory()
{
    //destory_2d(X, N);
    //destory_2d(Y, N);
    //destory_2d(field, N);
    //destory_1d(XX);
    //destory_1d(YY);
    cout << "Beam的destory" << endl;
}

Beam Beam::translate(double distance)
{
    int flag = 2;
    double L = Furion_NS::Max(this->X[0], N) - Furion_NS::Min(this->X[0], N);
    
    if (flag == 1)
    {
        f_f_spa.Furion_fresnel_spatialh(this->X, this->Y, this->field, this->wavelength, distance, this->N);
    }
    else if (flag == 2)
    {
        f_f_spe.Furion_fresnel_spectralh(this->N, 1 / L, this->field, this->wavelength, distance);

    }
    else if (flag == 3)
    {

    }
    else
    {

    }

    //for (int i = 0; i < N; i++)
    //{
    //    for (int j = 0; j < N; j++)
    //    {
    //        cout << this->field[i][j] << "  ";
    //    }
    //    cout << endl;
    //}

    return Beam(this->X, this->Y, this->field, this->wavelength, this->N);
}

Beam Beam::resize1(double L)
{
    //double* x = new double[this->N];
    //double* y = new double[this->N];
    std::vector<double> x(this->N);
    std::vector<double> y(this->N);
    Furion_NS::linspace(x, -L / 2, L / 2, this->N);
    Furion_NS::linspace(y, -L / 2, L / 2, this->N);

    //std::complex<double>** Field_new;
    //create_2d(Field_new, this->N);
    std::vector<std::vector<std::complex<double>> > Field_new(this->N, std::vector<std::complex<double>>(this->N));
    //scatteredInterpolant_2d_complex(Field_new, x, y, this->X, this->Y, this->field, this->N);

    //double** X_r;
    //double** Y_r;
    //create_2d(X_r, this->N);
    //create_2d(Y_r, this->N);
    std::vector<std::vector<double> > X_r(this->N, std::vector<double>(this->N));
    std::vector<std::vector<double> > Y_r(this->N, std::vector<double>(this->N));

    for (int i = 0; i < this->N; i++)
    {
        for (int j = 0; j < this->N; j++)
        {
            X_r[i][j] = x[j];
            Y_r[i][j] = y[i];
        }
    }
    
    return Beam(X_r, Y_r, Field_new, this->wavelength, this->N);
    cout << "Beam的resize1" << endl;
}

Beam Beam::resize2(double L, int N)
{
    std::vector<double> x(N);
    std::vector<double> y(N);
    Furion_NS::linspace(x, -L / 2, L / 2, N);
    Furion_NS::linspace(y, -L / 2, L / 2, N);

    std::vector<std::vector<std::complex<double>> > Field_new(N, std::vector<std::complex<double>>(N));
    //scatteredInterpolant_2d_complex(Field_new, x, y, this->X, this->Y, this->field, this->N);

    std::vector<std::vector<double> > X_r(N, std::vector<double>(N));
    std::vector<std::vector<double> > Y_r(N, std::vector<double>(N));

    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            X_r[i][j] = x[j];
            Y_r[i][j] = y[i];
        }
    }

    return Beam(X_r, Y_r, Field_new, this->wavelength, N);
    cout << "Beam的resize2" << endl;
}

Beam Beam::resize3(double Lx, double Ly, int N)
{
    std::vector<double> x(N);
    std::vector<double> y(N);
    Furion_NS::linspace(x, -Lx / 2, Lx / 2, N);
    Furion_NS::linspace(y, -Ly / 2, Ly / 2, N);

    std::vector<std::vector<std::complex<double>> > Field_new(N, std::vector<std::complex<double>>(N));
    //scatteredInterpolant_2d_complex(Field_new, x, y, this->X, this->Y, this->field, this->N);

    std::vector<std::vector<double> > X_r(N, std::vector<double>(N));
    std::vector<std::vector<double> > Y_r(N, std::vector<double>(N));

    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            X_r[i][j] = x[j];
            Y_r[i][j] = y[i];
        }
    }

    return Beam(X_r, Y_r, Field_new, this->wavelength, N);
    cout << "Beam的resize3" << endl;
}

double Beam::power()
{
    int N = this->X[0].size();
    double p = 0;
    double cache;

    std::vector<double> data(N);
    std::vector<double> y(N);
    std::vector<std::vector<double> > field2(N, std::vector<double>(N));
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            cache = std::abs(this->field[i][j]);
            field2[i][j] = cache * cache;
        }
        y[i] = this->Y[i][0];
    }
    
    trapz(data, this->X[0], field2, 2.0);
    trapz(p, y, data, 1);

    return p;
}

void Beam::trapz(std::vector<double>& out, std::vector<double>& x, std::vector<std::vector<double>>& y, double dx) 
{
    int num_points = x.size();
    int num_cols = y[0].size();
    double sum;

    for (int j = 0; j < num_cols; ++j) 
    {
        sum = 0.0;
        for (int i = 1; i < num_points; ++i) 
        {
            sum += (y[i - 1][j] + y[i][j]) * (x[i] - x[i - 1]) / dx;
        }
        out[j] = sum;
    }
}

void Beam::trapz(double& out, std::vector<double>& x, std::vector<double>& y, double dx)
{
    int num_points = x.size();
    out = 0;

    for (int i = 1; i < num_points; ++i) 
    {
        out += (y[i - 1] + y[i]) * (x[i] - x[i - 1]) / dx;
    }
}